### PR TITLE
Journaling local store operations

### PR DESCRIPTION
**Pull request by shrishrirang**
- Insert / Update / Delete operations performed on the local store using the sync context will be logged to the operation table.
- The operationTableManager module implements the operation table functionality
- MobileServiceSyncContext module has been updated to log all write operations

### CODE REVIEW

**Comment by shrishrirang** [comment]
\+ @danderson00

File: sdk/src/sync/MobileServiceSyncContext.js [hunk 1]
```diff
@@ -52,25 +57,36 @@ function MobileServiceSyncContext(client) {
...
             Validate.notNullOrEmpty(tableName, 'tableName');
 
             Validate.notNull(instance, 'instance');
             Validate.isValidId(instance.id, 'instance.id'); //TODO(shrirs): Generate an ID if ID is not defined
-        }).then(function() {
-            return store.lookup(tableName, instance.id);
-        }).then(function(result) {
-            if (!_.isNull(result)) {
-                throw new Error('Cannot perform insert as a record with ID ' + id + ' already exists in the table ' + tableName);
+
```
Comment (danderson00) [inline sdk/src/sync/MobileServiceSyncContext.js]: This block seems mostly duplicated for insert, update and delete operations. Refactor out into a single function with parameters? Keeps file shorter and slightly easier to understand.

**Comment by danderson00** [comment]
What's the locking stuff about? Might be better to explain face to face.

**Comment by shrishrirang** [comment]
We lock operations while a push is in progress. This is so that we don't condense other operations into this operation if we have already begun a push for it.

---
In reply to: [220097088](https://github.com/Azure/azure-mobile-apps-js-client/pull/113#issuecomment-220097088) [](ancestors = 220097088)

**Comment by danderson00** [comment]
Looks good, pending that minor refactor

**Pull request merged**